transformer_engine/pytorch/optimizers/fused_adam.py

```diff
@@ -172,45 +211,170 @@ def step(self, closure=None, grad_scaler=None):
-
-                state = self.state[p]
-                # State initialization
-                if len(state) == 0:
-                    # Exponential moving average of gradient values
-                    state["exp_avg"] = torch.zeros_like(p.data).float()
-                    # Exponential moving average of squared gradient values
-                    state["exp_avg_sq"] = torch.zeros_like(p.data).float()
-
-                if p.dtype == torch.float16:
-                    if self.master_weights:
-                        p_16_master.append(p_master.data)
-                    g_16.append(p.grad.data)
-                    p_16.append(p.data)
-                    m_16.append(state["exp_avg"])
-                    v_16.append(state["exp_avg_sq"])
-                elif p.dtype == torch.bfloat16:
-                    g_bf.append(p.grad)
-                    p_bf.append(p)
-                    m_bf.append(state["exp_avg"])
-                    v_bf.append(state["exp_avg_sq"])
-                elif p.dtype == torch.float32:
-                    if self.master_weights:
-                        p_32_master.append(p_master.data)
-                    g_32.append(p.grad.data)
-                    p_32.append(p.data)
-                    m_32.append(state["exp_avg"])
-                    v_32.append(state["exp_avg_sq"])
-                else:
-                    raise RuntimeError("FusedAdam only support fp16 and fp32.")
+            # when fuse_dtype_casting is True, the adam kernel also needs to update extra parameters.
+            p_main_of_fp8 = []
+            p_main_of_f16 = []
+            g_main_of_fp8 = []
+            g_main_of_f16 = []
+            m_main_of_fp8 = []
+            m_main_of_f16 = []
+            v_main_of_fp8 = []
+            v_main_of_f16 = []
+            p_extra_fp16_out = []
+            p_extra_fp8_out = []
+            scales = []
+            amaxes = []
+
+            # Only used when extra params include fp8 tensors. Otherwise, it doesn't matter what the out_dtype is.
+            out_dtype = tex.DType.kFloat32
+
+            # check extra params if fuse_dtype_casting is True
+            if self.fuse_dtype_casting:
+                assert (
+                    extra_group is not None
+                ), "extra_group should not be None when fuse_dtype_casting is True."
+                same_cnt = 0
+                cnt = 0
+                for p, p_extra in zip(group["params"], extra_group):
+                    assert (
+                        p_extra is not None and p_extra.shape == p.shape
+                    ), "Extra parameter is None or has different shape."
+                    cnt += 1
+                    if p.data is p_extra.data:
+                        cnt += 1
+                assert same_cnt in (cnt, 0), (
+                    "Either all extra params are identical to the main params, or they are all"
+                    " different."
+                )
```

Comment:
This is fine, but also not necessary for correctness. There will just be some unnecessary writes if some of the extra params are identical to the main params.